app/scripts/shell/dimMaterialsExchangeCtrl.controller.js

```diff
@@ -0,0 +1,197 @@
+(function() {
+  'use strict';
+
+  angular.module('dimApp').controller('dimMaterialsExchangeCtrl', MaterialsController);
+
+  MaterialsController.$inject = ['$scope', 'dimItemService', 'dimStoreService', '$state', 'dimFeatureFlags'];
+
+  function MaterialsController($scope, dimItemService, dimStoreService, $state, dimFeatureFlags) {
+    if (!dimFeatureFlags.materialsExchangeEnabled) {
+      $state.go('inventory');
+      return;
+    }
+
+    var vm = this;
+    vm.repPool = {};
+    vm.newRank = 0;
+    vm.newExperiance = 0;
+
+    vm.factions = [
+      // 174528503, eris breaks things atm
+      3871980777,
+      1424722124,
+      2778795080
+    ];
+
+    vm.characters = dimStoreService.getStores()
+      .filter(function(character) {
+        return !character.isVault;
+      })
+      .map(function(item) {
+        var background = item.background;
+        var icon = item.icon;
+        var className = item.className;
+        var genderRace = item.genderRace;
+        var factions = item.progression.progressions.filter((faction) => {
+          return vm.factions.includes(faction.hash);
+        }).reduce(function(map, obj) {
+          map[obj.hash] = obj;
+          return map;
+        }, {});
+        return { background, icon, className, genderRace, factions };
+      });
+
+    if (vm.characters.length === 0) {
+      $state.go('inventory');
+      return;
+    }
+
+    vm.activeCharacter = vm.characters[0];
+    vm.activeFaction = vm.activeCharacter.factions[1424722124];
+    vm.setActiveFaction = function(hash) {
+      vm.activeFaction = vm.activeCharacter.factions[hash];
+      updateRanks();
+    };
+
+    vm.setActiveCharacter = function(character) {
+      vm.activeCharacter = character;
+      vm.setActiveFaction(vm.activeFaction.hash);
+    };
+
+    vm.toggleItem = function(item) {
+      if (vm.repPool[item.hash]) {
+        delete vm.repPool[item.hash];
+      } else {
+        vm.repPool[item.hash] = vm.calculateRep(item);
+      }
+      updateRanks();
+    };
+
+    function updateRanks() {
+      // rank 1 = 1500
+      // rank 2 = 2000
+      // rank 3+ = 2500
+      var total = 0;
+
+      for (var key in vm.repPool) {
+        total = total + vm.repPool[key];
+      }
+
+      var totalRank = total + vm.activeFaction.progressToNextLevel;
+      vm.newRank = Math.floor((totalRank) / 2500) + vm.activeFaction.level;
+      vm.newExperiance = (totalRank % 2500);
+    }
+
+
+    var materialsHashes = [
```

Comment:
You can go through the vendors objects now to get most of this info - it should be possible to generate this page pretty much automatically, without any hardcoding.